File: 290-WordPattern/290-WordPattern.java
/*

290. Word Pattern
Easy

Given a pattern and a string str, find if str follows the same pattern.

Here follow means a full match, such that there is a bijection between a letter in pattern and a non-empty word in str.

Example 1:

Input: pattern = "abba", str = "dog cat cat dog"
Output: true

Example 2:

Input:pattern = "abba", str = "dog cat cat fish"
Output: false

Example 3:

Input: pattern = "aaaa", str = "dog cat cat dog"
Output: false

Example 4:

Input: pattern = "abba", str = "dog dog dog dog"
Output: false

Notes:
You may assume pattern contains only lowercase letters, and str contains lowercase letters that may be separated by a single space.

*/

class Solution {
    public boolean wordPattern(String pattern, String str) {
        if(!validateInput(pattern, str)) return false;
        LinkedHashMap<Character, List<Integer>> map = new LinkedHashMap<>();
        for(int i = 0; i < pattern.length(); i++){
            char ch = pattern.charAt(i);
            if(map.containsKey(ch)) map.get(ch).add(i);
            else{
                List<Integer> list = new ArrayList<>();
                list.add(i);
                map.put(ch, list);
            }
        }
        String[] arr = str.split(" ");
        Set<String> vis = new HashSet<>();
        for(Map.Entry<Character, List<Integer>> entry : map.entrySet()){
            List<Integer> list = entry.getValue();
            for(int i = 0; i < list.size() - 1; i++){
                if(!arr[list.get(i)].equals(arr[list.get(i + 1)]) || vis.contains(arr[list.get(i)])) return false;
            }
            if(vis.contains(arr[list.get(0)])) return false;
            vis.add(arr[list.get(0)]);
        }
           return true;
    }
    private boolean validateInput(String pattern, String str) {
        if(pattern == null || str == null || pattern.length() != str.split(" ").length || pattern.length() > 0 && str.length() == 0) return false;
        return true;
    }
}

File: 290-WordPattern/290-WordPattern.py
"""

290. Word Pattern

Given a pattern and a string s, find if s follows the same pattern.

Here follow means a full match, such that there is a bijection between a letter in pattern and a non-empty word in s. Specifically:

    Each letter in pattern maps to exactly one unique word in s.
    Each unique word in s maps to exactly one letter in pattern.
    No two letters map to the same word, and no two words map to the same letter.

Example 1:

Input: pattern = "abba", s = "dog cat cat dog"

Output: true

Explanation:

The bijection can be established as:

    'a' maps to "dog".
    'b' maps to "cat".

Example 2:

Input: pattern = "abba", s = "dog cat cat fish"

Output: false

Example 3:

Input: pattern = "aaaa", s = "dog cat cat dog"

Output: false

Constraints:

    1 <= pattern.length <= 300
    pattern contains only lower-case English letters.
    1 <= s.length <= 3000
    s contains only lowercase English letters and spaces ' '.
    s does not contain any leading or trailing spaces.
    All the words in s are separated by a single space.

"""

class Solution:
    def wordPattern(self, pattern: str, s: str) -> bool:
        smapp = defaultdict(list)
        pmapp = defaultdict(list)
        for i, ch in enumerate(pattern): pmapp[ch].append(i)
        for i, word in enumerate(s.split(" ")):  smapp[word].append(i)
        if len(smapp.keys()) != len(pmapp.keys()): return False
        for sk, pk in zip(smapp.keys(), pmapp.keys()):
            if smapp[sk] != pmapp[pk]: return False
        return True
